File: scala/src/main/scala/kv/KVClient.scala
/**
 * Scala gRPC KV Client with mTLS
 *
 * Implements a gRPC client for the KV service with mutual TLS authentication.
 * Uses grpc-java with Netty for TLS support.
 */
package kv

import com.google.protobuf.ByteString
import io.grpc._
import io.grpc.netty.{GrpcSslContexts, NettyChannelBuilder}
import io.netty.handler.ssl.SslContextBuilder

import java.io.ByteArrayInputStream
import java.time.Instant
import scala.concurrent.{Await, ExecutionContext}
import scala.concurrent.duration._

object KVClient {
  implicit val ec: ExecutionContext = ExecutionContext.global

  def log(level: String, message: String): Unit = {
    val timestamp = Instant.now().toString
    System.err.println(s"$timestamp [$level]     $message")
  }

  def main(args: Array[String]): Unit = {
    log("INFO", "Starting gRPC KV Client (Scala)")

    // Load certificates from environment
    val serverCert = Option(System.getenv("PLUGIN_SERVER_CERT")).getOrElse {
      log("ERROR", "Missing required environment variable: PLUGIN_SERVER_CERT")
      System.exit(1)
      ""
    }
    val clientCert = Option(System.getenv("PLUGIN_CLIENT_CERT"))
    val clientKey = Option(System.getenv("PLUGIN_CLIENT_KEY"))

    log("INFO", "Loading certificates...")
    log("INFO", s"Server cert length: ${serverCert.length} bytes")
    log("INFO", s"Client cert length: ${clientCert.map(_.length).getOrElse(0)} bytes")
    log("INFO", s"Client key length: ${clientKey.map(_.length).getOrElse(0)} bytes")

    val host = Option(System.getenv("PLUGIN_HOST")).getOrElse("localhost")
    val port = Option(System.getenv("PLUGIN_PORT")).map(_.toInt).getOrElse(50051)
    val address = s"$host:$port"

    log("INFO", s"Connecting to $address")

    try {
      // Build SSL context
      var sslContextBuilder = SslContextBuilder.forClient()
        .trustManager(new ByteArrayInputStream(serverCert.getBytes("UTF-8")))

      (clientCert, clientKey) match {
        case (Some(cert), Some(key)) =>
          sslContextBuilder = sslContextBuilder.keyManager(
            new ByteArrayInputStream(cert.getBytes("UTF-8")),
            new ByteArrayInputStream(key.getBytes("UTF-8"))
          )
          log("INFO", "mTLS credentials configured")
        case _ =>
          log("INFO", "TLS credentials configured (no client auth)")
      }

      val sslContext = GrpcSslContexts.configure(sslContextBuilder).build()

      // Create channel
      val channel = NettyChannelBuilder.forAddress(host, port)
        .sslContext(sslContext)
        .overrideAuthority("localhost")
        .build()

      val client = KVGrpc.stub(channel)

      // Test Get operation
      log("INFO", "Sending Get request...")
      val getRequest = GetRequest(key = "test-key")
      val getResponse = Await.result(client.get(getRequest), 30.seconds)
      log("INFO", s"Get response: ${getResponse.value.toStringUtf8}")

      // Test Put operation
      log("INFO", "Sending Put request...")
      val putRequest = PutRequest(
        key = "test-key",
        value = ByteString.copyFromUtf8("test-value")
      )
      Await.result(client.put(putRequest), 30.seconds)
      log("INFO", "Put request successful")

      channel.shutdown()

      log("INFO", "All operations completed successfully")
      println("OK")
      System.exit(0)
    } catch {
      case e: Exception =>
        log("ERROR", s"Client error: ${e.getMessage}")
        e.printStackTrace()
        System.exit(1)
    }
  }
}

File: nodejs/kv-client.js
#!/usr/bin/env node

/**
 * Node.js gRPC KV Client with mTLS
 *
 * Connects to a KV server using mutual TLS authentication.
 * Matches the interface defined in proto/kv.proto
 */

const grpc = require('@grpc/grpc-js');
const protoLoader = require('@grpc/proto-loader');
const path = require('path');
const crypto = require('crypto');

// Load the protobuf definition
const PROTO_PATH = path.join(__dirname, '..', 'proto', 'kv.proto');
const packageDefinition = protoLoader.loadSync(PROTO_PATH, {
    keepCase: true,
    longs: String,
    enums: String,
    defaults: true,
    oneofs: true
});
const kvProto = grpc.loadPackageDefinition(packageDefinition).proto;

// Logger with structured output
function log(level, domain, action, status, message) {
    const timestamp = new Date().toISOString();
    console.log(`${timestamp} [${level}] ${domain} ${action} ${status} ${message}`);
}

// Certificate info logging
function logCertificateInfo(cert, prefix) {
    try {
        const x509 = new crypto.X509Certificate(cert);
        log('INFO', '  ', '  ', '  ', `${prefix} Certificate Details:`);
        log('INFO', '  ', '  ', '  ', `  Subject: ${x509.subject}`);
        log('INFO', '  ', '  ', '  ', `  Issuer: ${x509.issuer}`);
        log('INFO', '  ', '  ', '  ', `  Valid From: ${x509.validFrom}`);
        log('INFO', '  ', '  ', '  ', `  Valid To: ${x509.validTo}`);
        log('INFO', '  ', '  ', '  ', `  Serial: ${x509.serialNumber}`);

        // Log key info
        const keyType = x509.publicKey.asymmetricKeyType;
        if (keyType === 'ec') {
            const details = x509.publicKey.asymmetricKeyDetails;
            log('INFO', '  ', '  ', '  ', `  Public Key: EC ${details.namedCurve}`);
        } else if (keyType === 'rsa') {
            const details = x509.publicKey.asymmetricKeyDetails;
            log('INFO', '  ', '  ', '  ', `  Public Key: RSA ${details.modulusLength} bits`);
        }
    } catch (err) {
        log('WARN', '  ', '  ', '  ', `Failed to parse certificate: ${err.message}`);
    }
}

async function main() {
    log('INFO', '  ', '  ', '  ', 'Starting gRPC KV Client (Node.js)');

    // Load certificates from environment variables
    const clientCert = process.env.PLUGIN_CLIENT_CERT;
    const clientKey = process.env.PLUGIN_CLIENT_KEY;
    const serverCert = process.env.PLUGIN_SERVER_CERT;

    if (!clientCert || !clientKey) {
        log('ERROR', '  ', '  ', '  ', 'Missing required environment variables: PLUGIN_CLIENT_CERT, PLUGIN_CLIENT_KEY');
        process.exit(1);
    }

    log('INFO', '  ', '  ', '  ', 'Loading certificates...');
    log('INFO', '  ', '  ', '  ', `Client cert length: ${clientCert.length} bytes`);
    log('INFO', '  ', '  ', '  ', `Client key length: ${clientKey.length} bytes`);
    log('INFO', '  ', '  ', '  ', `Server cert length: ${serverCert ? serverCert.length : 0} bytes`);

    // Log certificate details
    logCertificateInfo(clientCert, 'Client');
    if (serverCert) {
        logCertificateInfo(serverCert, 'Server');
    }

    // Create client credentials with mTLS
    let credentials;
    try {
        if (serverCert) {
            // mTLS with known server certificate
            credentials = grpc.credentials.createSsl(
                Buffer.from(serverCert),  // Root CA for server verification
                Buffer.from(clientKey),    // Client private key
                Buffer.from(clientCert)    // Client certificate chain
            );
            log('INFO', '  ', '  ', '  ', 'mTLS credentials configured');
        } else {
            log('ERROR', '  ', '  ', '  ', 'Server certificate required for mTLS');
            process.exit(1);
        }
    } catch (err) {
        log('ERROR', '  ', '  ', '  ', `Failed to create credentials: ${err.message}`);
        process.exit(1);
    }

    // Channel options
    const options = {
        'grpc.ssl_target_name_override': 'localhost',
        'grpc.default_authority': 'localhost',
        'grpc.max_receive_message_length': 100 * 1024 * 1024,
        'grpc.max_send_message_length': 100 * 1024 * 1024,
        'grpc.keepalive_time_ms': 10000,
        'grpc.keepalive_timeout_ms': 5000,
        'grpc.keepalive_permit_without_calls': 1
    };

    // Connect to server
    const host = process.env.PLUGIN_HOST || 'localhost';
    const port = process.env.PLUGIN_PORT || '50051';
    const address = `${host}:${port}`;

    log('INFO', '  ', '  ', '  ', `Connecting to server at ${address}...`);

    const client = new kvProto.KV(address, credentials, options);

    // Wait for connection
    const deadline = new Date();
    deadline.setSeconds(deadline.getSeconds() + 10);

    await new Promise((resolve, reject) => {
        client.waitForReady(deadline, (err) => {
            if (err) {
                log('ERROR', '  ', '  ', '  ', `Failed to connect: ${err.message}`);
                reject(err);
            } else {
                log('INFO', '  ', '  ', '  ', 'Connected successfully');
                resolve();
            }
        });
    });

    // Send Get request
    log('INFO', '  ', '  ', '  ', 'Sending Get request...');

    const response = await new Promise((resolve, reject) => {
        client.Get({ key: 'test' }, (err, response) => {
            if (err) {
                log('ERROR', '  ', '  ', '  ', `Get request failed: ${err.message}`);
                log('ERROR', '  ', '  ', '  ', `  Code: ${err.code}`);
                log('ERROR', '  ', '  ', '  ', `  Details: ${err.details}`);
                reject(err);
            } else {
                resolve(response);
            }
        });
    });

    const value = response.value ? response.value.toString() : '';
    console.log(`Response: ${value}`);
    log('INFO', '  ', '  ', '  ', 'Request completed successfully');

    // Clean up
    client.close();
}

main().catch(err => {
    log('ERROR', '  ', '  ', '  ', `Fatal error: ${err.message}`);
    process.exit(1);
});
